File: tests/fixtures/dart/control-flow.dart
typedef Callback = String Function();

class Failure {
  String report() => "failure";
}

Callback make() => () => "ready";
bool ready() => true;
String fallback() => "fallback";
void work() {}

(Callback, Callback) pair() => (make(), make());

void control(
  List<Callback> callbacks,
  List<(Callback, Callback)> pairs,
  Object value,
) {
  for (var loop = make(); ready(); loop = make()) {
    loop();
  }
  loop();

  for (final item in callbacks) {
    item();
  }
  item();

  for (final (left, right) in pairs) {
    left();
    right();
  }
  left();

  try {
    work();
  } on Failure catch (error, stack) {
    error.report();
    stack.toString();
  }
  error.report();
  stack.toString();

  var (first, second) = pair();
  first();
  second();

  if (value case Callback selected) {
    selected();
  }
  selected();

  switch (value) {
    case Callback matched when ready():
      matched();
      break;
    default:
      break;
  }
  matched();

  final rendered = switch (value) {
    Callback chosen => chosen(),
    _ => fallback(),
  };
  chosen();
  print(rendered);
}

File: tests/fixtures/kotlin/control-flow.kt
package fixtures.control

typealias Callback = () -> Unit

class Failure {
  fun report() {}
}

fun make(): Callback = {}
fun pair(): Pair<Callback, Callback> = Pair(make(), make())
fun work() {}

fun control(
  callbacks: List<Callback>,
  pairs: List<Pair<Callback, Callback>>,
  value: Any,
) {
  for (item in callbacks) {
    item()
  }
  item()

  for ((left, right) in pairs) {
    left()
    right()
  }
  left()

  try {
    work()
  } catch (error: Failure) {
    error.report()
  }
  error.report()

  val (first, second) = pair()
  first()
  second()

  when (val selected = make()) {
    else -> selected()
  }
  selected()

  val anonymous = fun(parameter: Callback) {
    parameter()
  }
  anonymous()
}
